pixeltable/utils/parquet.py

```diff
@@ -0,0 +1,132 @@
+import json
+import PIL.Image
+import io
+import pyarrow.parquet as pq
+import pyarrow as pa
+import numpy as np
+from pathlib import Path
+from collections import deque
+
+from pixeltable.dataframe import DataFrame
+from typing import List, Tuple, Any, Dict
+
+from pixeltable.utils.transactional_folder import transactional_folder
+
+
+def _write_batch(value_batch : Dict[str, deque], schema : pa.Schema, output_path : Path) -> None:
+    pydict = {}
+    for field in schema:
+        if isinstance(field.type, pa.FixedShapeTensorType):
+            stacked_arr = np.stack(value_batch[field.name])
+            pydict[field.name] = pa.FixedShapeTensorArray.from_numpy_ndarray(stacked_arr)
+        else:
+            pydict[field.name] = value_batch[field.name]
+
+    tab = pa.Table.from_pydict(pydict, schema=schema)
+    pq.write_table(tab, output_path)
+
+
+def save_parquet(df : DataFrame, dest_path : Path, partition_size_bytes : int = 100_000_000):
+    """
+        Internal method to stream dataframe data to parquet format.
+        Does not materialize the dataset to memory.
+
+        It preserves pixeltable type metadata in a json file, which would otherwise 
+        not be available in the parquet format.
+
+        Images are stored inline in a compressed format in their parquet file.
+
+        NB. pixeltable types
+            - string ok
+            - image ok (inlined)
+            - video ok (path only)
+    
+        Args:
+            partition_size_bytes: maximum target size for each chunk. Default 100MB.
+        Returns:
+            path to the parquet directory
+
+    """
+    column_names = df.get_column_names()
+    column_types = df.get_column_types()
+
+    arrow_schema = pa.schema([pa.field(name, column_types[i].to_arrow_type()) for i, name in enumerate(column_names)])
+    # store the changes atomically
+    with transactional_folder(dest_path) as temp_path:
+        # dump metadata json file so we can inspect what was the source of the parquet file later on.
+        json.dump(df.as_dict(), (temp_path / '.pixeltable.json').open('w')) # pylint: disable=protected-access
+
+        batch_num = 0
+        current_value_batch : Dict[str, deque] = {k:deque() for k in column_names}
```

Comment:
It'll get fixed when we have the auto-formatter in place, but in the meantime, could you just change everything, incl. type hints, to be like 'a<no space>:<space>...'?